persistence/trees/trees.go

```diff
@@ -0,0 +1,365 @@
+// package trees maintains a set of sparse merkle trees
+// each backed by the KVStore interface. It offers an atomic
+// commit and rollback mechanism for interacting with
+// that core resource map of merkle trees.
+package trees
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/jackc/pgx/v5"
+
+	"github.com/pokt-network/pocket/persistence/indexer"
+	"github.com/pokt-network/pocket/persistence/kvstore"
+	"github.com/pokt-network/pocket/persistence/sql"
+	"github.com/pokt-network/pocket/shared/codec"
+	coreTypes "github.com/pokt-network/pocket/shared/core/types"
+	"github.com/pokt-network/pocket/shared/crypto"
+	"github.com/pokt-network/smt"
+)
+
+var merkleTreeToString = map[merkleTree]string{
+	appMerkleTree:      "app",
+	valMerkleTree:      "val",
+	fishMerkleTree:     "fish",
+	servicerMerkleTree: "servicer",
+
+	accountMerkleTree: "account",
+	poolMerkleTree:    "pool",
+
+	transactionsMerkleTree: "transactions",
+	paramsMerkleTree:       "params",
+	flagsMerkleTree:        "flags",
+}
+
+var actorTypeToMerkleTreeName = map[coreTypes.ActorType]merkleTree{
+	coreTypes.ActorType_ACTOR_TYPE_APP:      appMerkleTree,
+	coreTypes.ActorType_ACTOR_TYPE_VAL:      valMerkleTree,
+	coreTypes.ActorType_ACTOR_TYPE_FISH:     fishMerkleTree,
+	coreTypes.ActorType_ACTOR_TYPE_SERVICER: servicerMerkleTree,
+}
+
+var merkleTreeToActorTypeName = map[merkleTree]coreTypes.ActorType{
+	appMerkleTree:      coreTypes.ActorType_ACTOR_TYPE_APP,
+	valMerkleTree:      coreTypes.ActorType_ACTOR_TYPE_VAL,
+	fishMerkleTree:     coreTypes.ActorType_ACTOR_TYPE_FISH,
+	servicerMerkleTree: coreTypes.ActorType_ACTOR_TYPE_SERVICER,
+}
+
+type merkleTree float64
+
+// A list of Merkle Trees used to maintain the state hash.
+const (
+	// IMPORTANT: The order in which these trees are defined is important and strict. It implicitly
```

Comment:
Wdyt of changing the root hash calculation to be the root of an SMT with key value pairings as follows:
`stateRootTree.Update([]byte("app"), treeStore.merkleTrees["app"].Root())`

Then our state root would be `stateRootTree.Root()`

Reply:
I'd like to see what this looks like spiked out but at first glance nothing jumps out at me as incompatible.

Reply:
Resolving per the ticket opened in #834. 

**Optional NIT**: Adding a TODO/techdebt in the code